### PR TITLE
Addition of SPADE Network  + tests and modification of SPADE normalisation 

### PR DESCRIPTION
**Pull request by virginiafdez**
- spade_network, and SPADENet (VAE-GAN)
- test_spade_vaegan (to test previously mentioned model) Modification of:
- spade_diffusion_model_unet.py to change namings.
- SPADE normalisation layer, to use get_norm_layer function instead of defining such layers directly.

Fixes # .

### Description

A few sentences describing the changes proposed in this pull request.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.

### CODE REVIEW

File: tests/test_spade_vaegan.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            out, z_mu, z_logvar = net(in_label, in_image)
+            self.assertEqual(
+                False,
+                True in torch.isnan(out)
+                or True in torch.isinf(out)
+                or True in torch.isinf(z_mu)
+                or True in torch.isnan(z_mu)
+                or True in torch.isinf(z_logvar)
+                or True in torch.isnan(z_logvar),
+            )
```
Comment (ericspod) [inline tests/test_spade_vaegan.py:123]:
```suggestion
            self.assertTrue(torch.all(torch.isfinite(out)))
            self.assertTrue(torch.all(torch.isfinite(z_mu)))
            self.assertTrue(torch.all(torch.isfinite(z_logvar)))
```

Comment (ericspod) [reply tests/test_spade_vaegan.py:123]: Same thing for the similar statement in `test_forward_2d`.

**Comment by ericspod** [comment]
The quick tests are failing for some weird issue with Matplotlib that may go away if they're rerun anyway, it isn't related to this PR. The DCO can be fixed as well by following the remediation instructions.

**Comment by marksgraham** [comment]
This is the final piece of #7227 

The need for the addition of the spade network here is my fault - I missed it out in the initial port, and @virginiafdez spotted it.

The matplotlib errors have been popping up on other PRs too, recently. 

Will do a review in a moment. I can help with the mypy errors too

**Review by marksgraham — APPROVED** [review]
looks good to me

File: monai/networks/nets/spade_network.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+            spatial_dims=spatial_dims,
+            in_channels=channels[-1],
+            out_channels=out_channels,
+            padding=(kernel_size - 1) // 2,
+            kernel_size=kernel_size,
+            norm=None,
+            act=last_act,
+        )
+
+    def forward(self, seg, z: torch.Tensor | None = None):
```
Comment (KumoLiu) [inline monai/networks/nets/spade_network.py:299]: Can we add docstring here mention what z represents here? Looks like it's output from the encoder, and will random create one if not specified with `self.is_gan=False`. Correct me if I'm wrong. Thanks.

File: monai/networks/nets/spade_network.py [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+        if self.is_gan:
+            x = F.interpolate(seg, size=tuple(self.latent_spatial_shape))
+            x = self.fc(x)
+        else:
+            if z is None:
+                z = torch.randn(seg.size(0), self.opt.z_dim, dtype=torch.float32, device=seg.get_device())
```
Comment (KumoLiu) [inline monai/networks/nets/spade_network.py:305]: I didn't find where this `self.opt` define.

**Review by KumoLiu — COMMENTED** [review]
Thanks for the PR, overall looks good to me.
Minor comments inline.

**Comment by dzenanz** [comment]
Perhaps rename this PR to `SPADENet` or `SPADENet (VAE-GAN)`? The current name, `Addition of:`, is quite non-descriptive.

File: monai/networks/nets/spade_network.py (hunk 3 above)
Comment (virginiafdez) [reply monai/networks/nets/spade_network.py:305]: This is a bug! Thanks for spotting. Should be self.z_dim, which has to be set-up in the init method.

File: tests/test_spade_vaegan.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    # One hot encode label
+    out_label_ = torch.zeros([semantic_regions] + list(out_label.shape))
+    for ch in range(semantic_regions):
+        out_label_[ch, ...] = out_label == ch
+
+    return out_label_.unsqueeze(0), out_image.unsqueeze(0).unsqueeze(0)
+
+
+class TestDiffusionModelUNet2D(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (virginiafdez) [reply tests/test_spade_vaegan.py:72]: This shouldn't be named like this indeed. Name changed to TestSpadeNet.

File: tests/test_spade_vaegan.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+import numpy as np
+import torch
+from parameterized import parameterized
+
+from monai.networks import eval_mode
+from monai.networks.nets import SPADENet
+
+CASE_2D = [[[2, 1, 1, 3, [64, 64], [16, 32, 64, 128], 16, True]]]
+CASE_2D_BIS = [[[2, 1, 1, 3, [64, 64], [16, 32, 64, 128], 16, True]]]
```
[the remark this answers is not in the document]
Comment (virginiafdez) [reply tests/test_spade_vaegan.py:24]: None, CASE_2D could be used instead for the test in question. Removed it.

File: tests/test_spade_vaegan.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+CASE_3D = [[[3, 1, 1, 3, [64, 64, 64], [16, 32, 64, 128], 16, True]]]
```
[the remark this answers is not in the document]
Comment (virginiafdez) [reply tests/test_spade_vaegan.py:25]: Multiple configs can be place within each case, but in this scenario, the cases had only one config. Now there are two per case.

File: monai/networks/nets/spade_network.py [hunk 7]
```diff
@@ -297,12 +306,23 @@ def __init__(
...
+        if not self.is_vae:
             x = F.interpolate(seg, size=tuple(self.latent_spatial_shape))
-            x = self.fc(x)
+            x = self.conv_init(x)
         else:
-            if z is None:
-                z = torch.randn(seg.size(0), self.opt.z_dim, dtype=torch.float32, device=seg.get_device())
+            if (
+                z is None and self.z_dim is not None
+            ):  # is_vae is Truee, but we can use the VAE-GAN as GAN in this function.
```
Comment (KumoLiu) [inline monai/networks/nets/spade_network.py:324]:
```suggestion
            ):  # is_vae is True, but we can use the VAE-GAN as GAN in this function.
```

I didn't get the point of this line of comment.

File: tests/test_spade_vaegan.py [hunk 8]
```diff
@@ -69,7 +74,7 @@ def create_semantic_data(shape: list, semantic_regions: int):
     return out_label_.unsqueeze(0), out_image.unsqueeze(0).unsqueeze(0)
 
 
-class TestDiffusionModelUNet2D(unittest.TestCase):
+class TestSpadeNet(unittest.TestCase):
```
Comment (KumoLiu) [inline tests/test_spade_vaegan.py:77]: Do we need to add a `load_old_state_dict` function for this network?
cc @ericspod @marksgraham

Comment (ericspod) [reply tests/test_spade_vaegan.py:77]: This being a test network we can regenerate weights for it without worrying about user compatibility.

Comment (ericspod) [reply tests/test_spade_vaegan.py:77]: We decided on not doing a old state dict loader for this network considering its use.

File: monai/networks/nets/spade_network.py (hunk 7 above)
Comment (virginiafdez) [reply monai/networks/nets/spade_network.py]: When flag is_vae is True, the network is a VAE-GAN, but it should be possible to sample from random noise as well, even when the network is a VAE-GAN (you either encode an input and then decode the latent resulting of that, or sample from noise). I've clarified this.
